OpenRA.Mods.Common/Traits/BotModules/CratePickupBotModule.cs

```diff
@@ -0,0 +1,150 @@
+	{
+		[Desc("Actor types that should not start hunting for crates.")]
+		public readonly HashSet<string> ExcludedUnitTypes = new HashSet<string>();
+
+		[Desc("Only these actor types should start hunting for crates.")]
+		public readonly HashSet<string> IncludedUnitTypes = new HashSet<string>();
+
+		[Desc("Interval (in ticks) between giving out orders to idle units.")]
+		public readonly int ScanForCratesInterval = 50;
+
+		[Desc("Only move this far away from base. Disabled if set to zero.")]
+		public readonly int MaxProximityRadius = 0;
+
+		[Desc("Avoid enemy actors nearby when searching for crates. Should be somewhere near the max weapon range.")]
+		public readonly WDist EnemyAvoidanceRadius = WDist.FromCells(8);
+
+		[Desc("Should visibility (Shroud, Fog, Cloak, etc) be considered when searching for crates?")]
+		public readonly bool CheckTargetsForVisibility = true;
+
+		public override object Create(ActorInitializer init) { return new CratePickupBotModule(init.Self, this); }
+	}
+
+	public class CratePickupBotModule : ConditionalTrait<CratePickupBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		readonly int maxProximity;
+
+		CrateSpawner crateSpawner;
+
+		IPathFinder pathfinder;
+		DomainIndex domainIndex;
+		int scanForCratesTicks;
+
+		List<Actor> alreadyPursuitCrates = new List<Actor>();
+
+		public CratePickupBotModule(Actor self, CratePickupBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			maxProximity = Info.MaxProximityRadius > 0 ? info.MaxProximityRadius : world.Map.Grid.MaximumTileSearchRange;
+		}
+
+		protected override void Created(Actor self)
+		{
+			crateSpawner = self.Owner.World.WorldActor.TraitOrDefault<CrateSpawner>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			pathfinder = world.WorldActor.Trait<IPathFinder>();
+			domainIndex = world.WorldActor.Trait<DomainIndex>();
+			scanForCratesTicks = Info.ScanForCratesInterval;
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (crateSpawner == null || !crateSpawner.IsTraitEnabled() || !crateSpawner.Enabled)
+				return;
+
+			if (--scanForCratesTicks > 0)
+				return;
+
+			scanForCratesTicks = Info.ScanForCratesInterval;
+
+			var crates = world.ActorsHavingTrait<Crate>().ToList();
+			if (!crates.Any())
+				return;
+
+			if (Info.CheckTargetsForVisibility)
+				crates.RemoveAll(c => !c.CanBeViewedByPlayer(player));
+
+			var idleUnits = world.ActorsHavingTrait<Mobile>().Where(a => a.Owner == player && a.IsIdle
+				&& (Info.IncludedUnitTypes.Contains(a.Info.Name) || (!Info.IncludedUnitTypes.Any() && !Info.ExcludedUnitTypes.Contains(a.Info.Name)))).ToList();
+
+			if (!idleUnits.Any())
+				return;
+
+			foreach (var crate in crates)
+			{
+				if (alreadyPursuitCrates.Contains(crate))
+					continue;
+
+				if (!crate.IsAtGroundLevel())
+					continue;
+
+				var crateCollector = idleUnits.ClosestTo(crate);
+				if (crateCollector == null)
+					return;
```

Comment:
Could you explain why you're returning rather than continuing on lines 112, 115, and 121?

Reply:
Well spotted. Fixing my mistake here greatly improves the AI performance.